src/dmtest/gendatablocks.py

```diff
@@ -0,0 +1,663 @@
+        raise CompareError("block not recognized.", block_number, actual)
+
+    def write(self,
+              tag: str,
+              dedupe: float = 0.0,
+              compress: float = 0.0,
+              direct: bool = False,
+              sync: bool = False,
+              fsync: bool = False):
+        """Write to a block range
+
+        Parameters
+        ----------
+        tag : str
+            tag the data for future reference
+        dedupe : float
+            how much dedupe to write
+        compress : float
+            how much compression to write
+        direct : bool
+            open the device with O_DIRECT
+        sync : bool
+            open the device with O_SYNC
+        fsync : bool
+            write the device with fsync
+
+        Raises
+        ------
+        ValueError
+
+        """
+        if tag is None:
+            raise ValueError("tag is not defined")
+        if len(tag) >= MAX_TAG_SIZE:
+            raise ValueError("tag must be 8 characters or less")
+        mod_dedupe = numpy.rint(DEDUPE_MODULUS * dedupe)
+        if (mod_dedupe < 0) or (mod_dedupe > DEDUPE_MODULUS):
+            raise ValueError("the dedupe fraction " + str(dedupe)
+                             + " is invalid")
+        # This attempts to handle the space used by the header in all blocks
+        if (compress < 0.0) or (compress > 0.96):
+            raise ValueError("the compression fraction " + str(compress)
+                             + " is invalid")
+        stream = BlockStream(tag, mod_dedupe, compress, direct, sync, fsync)
+        self.write_stream(stream)
+
+    def write_stream(self, stream: 'DataStream'):
+        """write a stream to a block range
+
+        Parameters
+        ----------
+        stream : 'DataStream'
+            the type of stream to write
+
+        """
+        if self.path:
+            self.write_stream_to_file(stream)
+        else:
+            self.write_stream_to_files(stream)
+        self.streams.append(stream)
+
+    def write_stream_to_file(self, stream: 'DataStream'):
+        """write a stream to a file
+
+        Parameters
+        ----------
+        stream : 'DataStream'
+            The stream to write with
+
+        Raises
+        ------
+        ValueError
+        """
+        if self.path is None:
+            raise ValueError("the file/device path is invalid")
+
+        flags = os.O_WRONLY
+        if stream.sync:
+            flags |= os.O_SYNC
+        if self.create:
+            flags |= os.O_CREAT
+
+        with os.fdopen(os.open(self.path, flags), "r+b") as fd:
+            self._seek(fd)
+            for n in range(0, self.block_count):
+                data = stream.generate(n, self.block_size)
+                fd.write(data)
+                stream.counter += 1
+            if stream.fsync:
+                os.fsync(fd)
+
+    def write_stream_to_files(self, stream: 'DataStream'):
```

Comment:
Can we drop the many-files version for now and just write (or verify) one file/device?